common/core/overlaylayer.cpp

```diff
@@ -229,55 +229,45 @@ void OverlayLayer::InitializeState(HwcLayer* layer,
 
   int32_t left_constraint = layer->GetLeftConstraint();
   int32_t right_constraint = layer->GetRightConstraint();
-  if (left_constraint >= 0 && right_constraint >= 0) {
-    if (display_frame_.right > right_constraint) {
-      display_frame_.right = right_constraint;
-    }
+  int32_t left_source_constraint = layer->GetLeftSourceConstraint();
+  int32_t right_source_constraint = layer->GetRightSourceConstraint();
+  int32_t display_frame_left = display_frame_.left;
+  uint32_t frame_width = display_frame_.right - display_frame_.left;
+  uint32_t source_width = source_crop_.right - source_crop_.left;
+  uint32_t frame_offset_left = 0;
+  uint32_t frame_offset_right = frame_width;
 
-    if (display_frame_.left < left_constraint) {
-      display_frame_.left = left_constraint;
-    }
-
-    if (display_frame_.right < right_constraint) {
-      display_frame_.right =
-          std::max(display_frame_.left, display_frame_.right);
-    }
+  if (left_constraint >= 0 && right_constraint >= 0) {
+    if (display_frame_.right < left_source_constraint ||
+        display_frame_.left > right_source_constraint) {
+      display_frame_.left = 0;
```

Comment:
You can replace this with ignore flag in next branch.

Also check with logical enabled (2 displays) and mosaic for these 2 display enabled and check things work fine.